### PR TITLE
feat(sdk/csharp): implement Bindu C# SDK with gRPC integration, tests and polish

### PR DESCRIPTION
**Pull request by EmmyAkine**
## Summary

- **Problem**: Bindu could turn agents written in TypeScript and Kotlin into A2A-compliant
  microservices, but .NET/C# developers had no supported SDK, so a large ecosystem
  (Semantic Kernel, Microsoft.Extensions.AI, plain C# agents) had no path to Bindu.
- **Why it matters**: A C# SDK gives .NET agents identical DID identity, auth, x402
  payments, and A2A protocol support against the same Bindu core as the TS/Kotlin SDKs,
  keeping agent identity and protocol consistent across languages.
- **What changed**: Added `sdks/csharp/` — `Bindu.Sdk` with a one-call `Bindufy()` API
  (core discovery/launch, gRPC registration, callback server, 30s heartbeat, clean
  shutdown), the shared `agent_handler.proto` gRPC contract, full XML docs, a README, and
  a 31-test xUnit suite that runs in-process against a fake core (no Python/network needed).
- **What did NOT change** (scope boundary): The Python Bindu core, the TypeScript/Kotlin
  SDKs, and all existing code were untouched (25 new files, all under `sdks/csharp/`).
  Streaming (`HandleMessagesStream`) is intentionally not implemented — matches the TS/Kotlin
  SDKs. The `Skills` config field is not yet transmitted during registration.

## Change Type (select all that apply)

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [x] Documentation
- [ ] Security hardening
- [x] Tests
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Server / API endpoints
- [ ] Extensions (DID, x402, etc.)
- [ ] Storage backends
- [ ] Scheduler backends
- [ ] Observability / monitoring
- [ ] Authentication / authorization
- [ ] CLI / utilities
- [x] Tests
- [x] Documentation
- [ ] CI/CD / infra

> Note: this PR adds a new area not listed above — `sdks/csharp/` (new language SDK).

## Linked Issue/PR

- Closes #
- Related #

## User-Visible / Behavior Changes
New SDK only — no changes to existing users (for them: `None`).

- .NET developers can now write an agent in plain C# and call `Bindufy(config, handler)`
  to get DID, auth, x402, and A2A protocol support.
- SDK defaults: callback port auto-picked (`GrpcCallbackPort = 0`, falls back to a free
  port if busy), `DeploymentUrl = http://localhost:3773`, `ExposeDeployment = false`,
  `Version = 0.1.0`.

## Security Impact (required)

- New permissions/capabilities? (`No`)
- Secrets/credentials handling changed? (`No`)
- New/changed network calls? (`Yes`) — the SDK launches the Bindu core as a local child
  process and talks to it over localhost gRPC (register/heartbeat/unregister on :3774),
  hosts a local gRPC callback server (dynamic port) for the core to deliver tasks, and the
  core exposes the A2A server on :3773. All on the local machine by default.
- Database schema/migration changes? (`No`)
- Authentication/authorization changes? (`No`)
- **If any `Yes`, explain risk + mitigation**: The SDK's callback gRPC server binds
  `IPAddress.Any` (all interfaces), not just loopback. The port is dynamic/ephemeral and
  the address is only shared with the locally launched core. Mitigation: consider
  restricting the bind to loopback in a follow-up before broader release.

### Environment

- OS: Windows
- .NET SDK: 10.0.302
- Storage backend: N/A (tests use an in-process fake core)
- Scheduler backend: N/A

### Steps to Test

1. `cd sdks/csharp`
2. `dotnet test`
3. Observe all tests pass
4. Alternatively, there is another test process described in the readme which involves creating a new console test app that echos back the users prompt.

### Expected Behavior
- All 31 xUnit tests pass (registration, heartbeat, gRPC client/server, core launcher,
  agent lifecycle, response mapping, fake-core integration).

>For the alternate test process
- End-to-end: the echo test app returns the user's prompt with an echo message appended,
  and the A2A port responds to `messages/send` / `messages/get`.

### Actual Behavior

- `Passed! - Failed: 0, Passed: 31, Skipped: 0, Total: 31, Duration: 54 s`

## Evidence (attach at least one)

- [x] Test output / logs (`dotnet test` — 31/31 passed on the squashed branch)
Screenshot 2026-08-13 022738

- [ ] Failing test before + passing after (N/A — new code, no prior tests existed)
- [x] Screenrecording (Postman `messages/send` + `task/get` exchanges from the echo test app are available)
https://github.com/user-attachments/assets/f0f95708-0285-48be-8a2a-ac2251bf037f

## Human Verification (required)

What you personally verified (not just CI):
- **Verified scenarios**: Full unit + in-process integration suite runs green on the
  squashed commit (`dotnet test`: 31 passed, 0 failed); clean build of `Bindu.Sdk` and the
  test project on .NET 10. Full end-to-end loop verified with a small ConsoleApp echo test
  app (launches the real Bindu core, registers the agent, returns the user's prompt with an
  echo message appended).
- **Edge cases checked**: callback port in use → automatic fallback to a free port; core
  startup timeout handling; heartbeat failure logging; graceful shutdown paths.
- **What you did NOT verify**: NuGet packaging/publishing.

## Compatibility / Migration

- Backward compatible? (`Yes`) — additive only; no existing files modified.
- Config/env changes? (`No`) — new SDK requires the .NET 10 SDK and (at runtime) the
  Bindu Python core, as documented in the SDK README.
- Database migration needed? (`No`)

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: revert the single commit or delete
  `sdks/csharp/` — nothing else depends on it.
- Files/config to restore: none (no existing files were touched).
- Known bad symptoms reviewers should watch for: none expected — isolated new code.

## Risks and Mitigations
- **Risk**: The callback gRPC server binds `IPAddress.Any` instead of loopback.
  - **Mitigation**: Dynamic ephemeral port; address only shared with the locally launched
    core; plan to restrict to loopback in a follow-up.
- **Risk**: `CoreLauncher` uses a fixed gRPC port (3774), so two SDK agents on one machine
  would contend for it.
  - **Mitigation**: Port map documented in the README; making the port configurable is
    follow-up work.

## Checklist

- [x] Tests pass — `dotnet test` (31/31). Note: `uv run pytest` is N/A — this PR changes no Python code.
- [ ] Pre-commit hooks pass (`uv run pre-commit run --all-files`) — run before opening; not executed in this session. 
>Note: `uv run pre-commit run --files $(git ls-files sdks/csharp/)` ran and passed
Screenshot 2026-08-13 021106

- [x] Documentation updated — SDK README ships with the PR.
- [x] Security impact assessed
- [x] Human verification completed
- [x] Backward compatibility considered

## Summary by CodeRabbit

* **New Features**
  * Added a C# SDK for creating, configuring, registering, and managing Bindu agents.
  * Supports synchronous and asynchronous cleanup, automatic core startup, health reporting, capabilities, and heartbeat monitoring.
  * Added structured responses with content, state transitions, prompts, and metadata.
  * Added gRPC communication for agent registration, task handling, and unregistration.
* **Documentation**
  * Added comprehensive C# SDK setup, usage, API, lifecycle, troubleshooting, and testing guidance.
* **Tests**
  * Added broad coverage for registration, messaging, lifecycle management, heartbeats, responses, networking, and error handling.

### CODE REVIEW

File: sdks/csharp/src/GrpcServer.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        /// <summary>Stops and disposes the underlying web application.</summary>
+        public async void StopServerAsync() {
+            if (_app is not null) {
+                await _app.StopAsync();
+                await _app.DisposeAsync();
+                _app = null;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (EmmyAkine) [reply sdks/csharp/src/GrpcServer.cs:82]: CleanUp is specifically designed to run as long as the parent process exit. Process exit often don't wait for asynchronous code to finish executing before exiting. Force kills wouldn't even trigger the synchronous cleanup code. But regardless clean up still needs to be implemented and so it makes sense to make the async code block execution till execution is complete so all cleanup logics run before exit
